pennylane/decomposition/decomposition_graph.py

```diff
@@ -55,6 +57,69 @@
 from .utils import translate_op_alias
 
 
+@dataclass(frozen=True)
+class _OperatorNode:
+    """A node that represents an operator type."""
+
+    op: CompressedResourceOp
+    """The resource rep of the operator."""
+
+    num_work_wire_not_available: int
+    """The number of work wires NOT available to the decomposition of this operator
+
+    The choice of decomposition rule for an operator depends on how many work wires is available,
+    which can be can be calculated by subtracting this value from the total number of work wires.
+    This convention allows the graph to be constructed without specifying the total number of work
+    wires, so that the same graph can be solved multiple times with different values for the max
+    number of work wires.
+
+    """
+
+    work_wire_dependent: bool = False
+    """Whether the decomposition of this operator depend on work wires.
+
+    This is true if any of the decomposition rules for this operator directly uses work wires or
+    if any of the operators produced down the line has decomposition rules that use work wires.
+
+    """
+
+    def __hash__(self) -> int:
+        # If the decomposition of an operator does not depend on the availability of work wires
+        # at all, we don't need to have multiple nodes representing the same operator with
+        # different work wire budgets. Therefore, we override the __hash__ and __eq__ of a node
+        # so that the num_work_wires_not_available is taken into account only when the operator
+        # depends on work wires. Since we keep track of all existing operator nodes in a set,
+        # this allows us to quickly find an existing operator node for another instance of the
+        # same operator with a different work wire budget that doesn't ultimately matter.
+        if self.work_wire_dependent:
+            return hash((self.op, self.num_work_wire_not_available))
+        return hash(self.op)
+
+    def __eq__(self, other) -> bool:
+        return hash(self) == hash(other)
```

Comment:
Why not just compare the properties?

Reply:
Think I'd prefer to just compare the properties themselves, as hash isn't necessarily unique.  Rare chance, but you can end up with two different things having the same hash.

Reply:
> Rare chance, but you can end up with two different things having the same hash.
That's the entire point. I want two operators with different values for `num_work_wires_not_available` to have the same hash and be considered identical if `work_wire_dependent=False`

Reply:
Updated